src/coreclr/jit/morph.cpp

```diff
@@ -561,6 +561,13 @@ GenTree* Compiler::fgMorphExpandCast(GenTreeCast* tree)
         // we fix this by copying the GC pointer to a non-gc pointer temp.
         noway_assert(!varTypeIsGC(dstType) && "How can we have a cast to a GCRef here?");
 
+        // we can just change the type for constant addresses
+        if (oper->IsCnsIntOrI())
+        {
+            oper->gtType = TYP_I_IMPL;
+            return fgMorphTree(gtNewCastNode(TYP_I_IMPL, oper, false, dstType));
```

Comment:
This might create e.g. a nongc icon handle typed as TYP_I_IMPL. I think we have a couple of asserts expecting those to be always TYP_REF. What exactly this changes improves?

Reply:
1. The cast here seems to strip handle-ness.
2. Would restricting it to `TYP_BYREF` make it okay?
3. Added codegen diff to the PR description.

Reply:
> 1. The cast here seems to strip handle-ness.

Which doesn't sound righ 

> 2. Would restricting it to `TYP_BYREF` make it okay?
> 3. Added codegen diff to the PR description.

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=584865&view=ms.vss-build-web.run-extensions-tab) are small and are regressions so does it make sense?

Reply:
> Which doesn't sound right e.g. - does it strip it for relocable (AOT) handles as well?

Made it retype the handles instead.

> [Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=584865&view=ms.vss-build-web.run-extensions-tab) are small and are regressions so does it make sense?

The only diff currently is from a GC hole that's fixed in #99138, I'd guess this won't have any diffs in the BCL or tests outside of #99011 since those don't generally use AsPointer on statics or RVA statics.

Reply:
It's nice that it helped you to find a gc hole! No diff in BCL means it's basically untested, it's probably fine now but may fail in future if we introduce byref icon handle? I think we should do changes like these only if they provide clear benefits and risks are understood